fix(cmd): add network validation in coffee cmd inputs

Anyitechs opened the pull request:
This PR validates the argument passed into the `-n` flag, and ensures only supported CLN networks are created in the `/Users/User/.coffee` directory, and closes #270 .

### Implementation
The current implementation consists of:

- An enum (`ClnNetwork`) with `Mainnet`, `Testnet`, `Signet`, and `Regtest` as variants
- A `ClnNetwork` implementation (`impl`) with a `from_str` function that `match`es on the `network` (`&str`) argument and returns a `ClnNetwork` variant, and  `to_str` function that `match`es on the `ClnNetwork` variant and returns the network `String`.
- The `network` function validates the `-n` argument.

### Current Behaviour
The validation works as expected, so passing in a random name as the network name will not create the directory but the command returns the same output as below regardless of the network name. Also, passing in a random network name and running `coffee -n remote list` for the first time ignores the invalid network name and creates a `/Users/ifeanyichukwu/.coffee/bitcoin` directory for example, but running the command again doesn't create a `../.coffee/bitcoin` directory again.

```
ifeanyichukwu@IfeanyiukwusMBP coffee % coffee -n remote list
● Plugins installed
╭───────────────────────────────────────────╮
│ ●   Language   Name   Enabled   Exec path 
├───────────────────────────────────────────┤
╰───────────────────────────────────────────╯
```

### How to test

- cd into the `coffee` directory
- Run `make install`
- Run `coffee -n remote list`

vincenzopalazzo commented on coffee_cmd/src/cmd.rs:
```diff
@@ -1,5 +1,6 @@
 //! Coffee command line arguments definition.
 use clap::{Parser, Subcommand};
+use coffee_lib::{error, errors::CoffeeError};
```
```suggestion
use coffee_lib::error;
use coffee_lib::errors::CoffeeError;
```

vincenzopalazzo commented on coffee_cmd/src/cmd.rs:
```diff
@@ -133,6 +134,36 @@ impl From<&RemoteAction> for coffee_core::RemoteAction {
+}
+
+impl ClnNetwork {
+    fn from_str(network: &str) -> Option<Self> {
+        
+        match network {
+            "mainnet" => Some(Self::Mainnet),
+            "testnet" => Some(Self::Testnet),
+            "signet" => Some(Self::Signet),
+            "regtest" => Some(Self::Regtest),
+            _ => None,
```
What about Liquid?

vincenzopalazzo replied:
implement the `From<&str>` trait, do not reivent rust

vincenzopalazzo commented on coffee_cmd/src/cmd.rs:
```diff
@@ -133,6 +134,36 @@ impl From<&RemoteAction> for coffee_core::RemoteAction {
+        }
+    }
+
+    fn to_str(network: Self) -> Option<String> {
+        match network {
+            Self::Mainnet => Some("mainnet".to_string()),
+            Self::Testnet => Some("testnet".to_string()),
+            Self::Signet => Some("signet".to_string()),
+            Self::Regtest => Some("regtest".to_string()),
+        }
+    }
```
implement the `std::fmt::Display` trait not use this an hack

vincenzopalazzo commented on coffee_cmd/src/cmd.rs:
```diff
@@ -147,7 +178,10 @@ impl coffee_core::CoffeeArgs for CoffeeArgs {
 
     fn network(&self) -> Option<String> {
-        self.network.clone()
+        let network = self.network.clone().ok_or_else(|| error!("Network is not set")).ok()?;
+        let validated_network = ClnNetwork::from_str(&network.to_lowercase()).ok_or_else(|| error!("Invalid network name")).ok()?;
```
I think you do not run `make` command, this code looks like unformatted.

Anyitechs commented:
Thank you for the review. I'll work on these and update the PR.

vincenzopalazzo commented on coffee_cmd/src/cmd.rs:
```diff
@@ -133,6 +136,43 @@ impl From<&RemoteAction> for coffee_core::RemoteAction {
+            ClnNetwork::Regtest => "regtest",
+            ClnNetwork::Liquid => "liquid",
+        };
+        write!(f, "{}", s)
```
```suggestion
        write!(f, "{s}")
```

vincenzopalazzo commented on coffee_cmd/src/cmd.rs:
```diff
@@ -133,6 +136,43 @@ impl From<&RemoteAction> for coffee_core::RemoteAction {
+            "signet" => Ok(Self::Signet),
+            "regtest" => Ok(Self::Regtest),
+            "liquid" => Ok(Self::Liquid),
+            _ => Err(format!("{} is not a valid network name", network)),
```
```suggestion
            _ => Err(format!("{network} is not a valid network name")),
```

vincenzopalazzo commented on coffee_cmd/src/cmd.rs:
```diff
@@ -147,7 +187,17 @@ impl coffee_core::CoffeeArgs for CoffeeArgs {
+        let network = self
+            .network
+            .clone()
+            .ok_or_else(|| error!("Network is not set"))
```
```suggestion
            .ok_or_else(|| error!("Network is not defined"))
```

vincenzopalazzo commented on coffee_cmd/src/cmd.rs:
```diff
@@ -147,7 +187,17 @@ impl coffee_core::CoffeeArgs for CoffeeArgs {
+        let validated_network = ClnNetwork::try_from(network.to_lowercase()).ok();
+
+        match validated_network {
+            Some(valid_network) => format!("{}", valid_network).into(),
```
```suggestion
            Some(valid_network) => format!("{valid_network}").into(),
```

vincenzopalazzo reviewed (changes requested):
Good code, some small minor changes

Anyitechs commented:
> Good code, some small minor changes

Thank you for the review. I've addressed all the requested changes and updated the PR.

Let me know if there's anything I'm missing.

vincenzopalazzo commented on coffee_cmd/src/cmd.rs:
```diff
@@ -1,5 +1,8 @@
 //! Coffee command line arguments definition.
 use clap::{Parser, Subcommand};
+use coffee_lib::error;
+use coffee_lib::errors::CoffeeError;
+use std::fmt::Display;
```
look at how we organize the import, we follow the following rules

```
std import

external dependencies import

coffee importa

crate import
```

But this can be addressed in a followup PR

Anyitechs replied:
Thank you for pointing that out. I'll address that asap!

Anyitechs replied:
I've addressed this in #276.

vincenzopalazzo reviewed (approved):
ACK ddaa0dc1a51021d5fe37effe301f944c49c58049
